File: bratek20-architecture/src/main/java/pl/bratek20/architecture/properties/sources/yaml/YamlPropertiesSource.java
package pl.bratek20.architecture.properties.sources.yaml;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.type.TypeFactory;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import lombok.SneakyThrows;
import pl.bratek20.architecture.properties.api.PropertiesSource;
import pl.bratek20.architecture.properties.api.PropertiesSourceName;
import pl.bratek20.architecture.properties.api.PropertyKey;

import java.io.File;
import java.io.IOException;
import java.lang.reflect.Type;
import java.util.List;

public class YamlPropertiesSource implements PropertiesSource {
    @SneakyThrows
    public static <T> T parseYamlSection(String resourcePath, String yamlPath, TypeReference<T> typeReference) {
        ObjectMapper mapper = new ObjectMapper(new YAMLFactory());
        File file = new File(YamlPropertiesSource.class.getClassLoader().getResource(resourcePath).getFile());
        JsonNode rootNode = mapper.readTree(file);

        // Navigate to the specified YAML path
        String[] pathElements = yamlPath.split("\\.");
        JsonNode targetNode = rootNode;
        for (String element : pathElements) {
            targetNode = targetNode.path(element);
        }

        // Deserialize the target node into the specified type
        return mapper.convertValue(targetNode, typeReference);
    }

    @Override
    public PropertiesSourceName getName() {
        return new PropertiesSourceName("yaml");
    }

    @Override
    public <T> T get(PropertyKey key, Class<T> type) {
        var name = key.getValue();
        return parseYamlSection("application.yaml", name, new TypeReference<T>() {
            @Override
            public Type getType() {
                return type;
            }
        });
    }

    public static <T> TypeReference<List<T>> listTypeRef(Class<T> clazz) {
        return new TypeReference<>() {
            @Override
            public Type getType() {
                return TypeFactory.defaultInstance().constructCollectionType(List.class, clazz);
            }
        };
    }

    @Override
    public <T> List<T> getList(PropertyKey key, Class<T> clazz) {
        var name = key.getValue();
        return parseYamlSection("application.yaml", name, listTypeRef(clazz));
    }

    @Override
    public <T> boolean hasOfType(PropertyKey key, Class<T> type) {
        var name = key.getValue();
        ObjectMapper mapper = new ObjectMapper(new YAMLFactory());
        try {
            // Load the YAML file
            File file = new File(getClass().getClassLoader().getResource("application.yaml").getFile());
            JsonNode rootNode = mapper.readTree(file);

            // Navigate to the specified YAML path
            String[] pathElements = name.split("\\.");
            JsonNode targetNode = rootNode;
            for (String element : pathElements) {
                targetNode = targetNode.get(element);
                if (targetNode == null) {
                    return false; // Path does not exist
                }
            }

            // Attempt to parse the node into the specified type
            T parsedObject = mapper.treeToValue(targetNode, type);
            return parsedObject != null; // If parsing succeeds, the path exists and is of the specified type
        } catch (IOException | IllegalArgumentException e) {
            return false; // Parsing failed or path does not exist
        }
    }

}

File: bratek20-architecture/src/main/java/pl/bratek20/architecture/properties/sources/yaml/YamlPropertiesSource.kt
package pl.bratek20.architecture.properties.sources.yaml

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import pl.bratek20.architecture.context.api.ContextBuilder
import pl.bratek20.architecture.context.api.ContextModule
import pl.bratek20.architecture.exceptions.ApiException
import pl.bratek20.architecture.properties.api.PropertiesSource
import pl.bratek20.architecture.properties.api.PropertiesSourceName
import pl.bratek20.architecture.serialization.api.SerializationType
import pl.bratek20.architecture.serialization.api.SerializedValue
import java.io.File
import java.nio.file.Paths

class YamlPropertiesSource: PropertiesSource {
    constructor(propertiesPath: String) {
        this.propertiesPath = propertiesPath
    }

    var propertiesPath: String = ""
        set(value) {
            if (!value.endsWith(".yaml")) {
                throw ApiException("Yaml properties source path should end with .yaml extension")
            }
            field = value
        }

    private val mapper = ObjectMapper(YAMLFactory())
        .registerKotlinModule()
    private val jsonMapper = ObjectMapper()

    override fun getName(): PropertiesSourceName {
        return PropertiesSourceName(Paths.get(propertiesPath).toAbsolutePath().toString())
    }

    override fun getAllKeys(): Set<String> {
        return try {
            val file = getFile(propertiesPath)
            val rootNode = mapper.readTree(file)
            rootNode.fieldNames().asSequence().toSet()
        } catch (e: Exception) {
            emptySet()
        }
    }

    override fun getValue(keyName: String): SerializedValue {
        return try {
            val file = getFile(propertiesPath)
            val rootNode = mapper.readTree(file)
            val targetNode = navigateToYamlNode(rootNode, keyName)
            SerializedValue.create(
                value = jsonMapper.writeValueAsString(targetNode),
                type = SerializationType.JSON
            )
        } catch (e: Exception) {
            throw ApiException("Error while getting value for key: $keyName")
        }
    }

    private fun getFile(path: String): File {
        val nioPath = Paths.get(path)
        val file = nioPath.toFile()

        require(file.exists() && !file.isDirectory()) {
            "File does not exist or is dir for path: ${nioPath.toAbsolutePath()}"
        }
        return file
    }

    private fun navigateToYamlNode(rootNode: JsonNode, path: String): JsonNode? {
        val pathElements = path.split(".")
        var targetNode = rootNode
        for (element in pathElements) {
            targetNode = targetNode[element]
        }
        return targetNode
    }
}

class YamlPropertiesSourceImpl(
    private val propertiesPath: String = "properties.yaml"
) : ContextModule {
    override fun apply(builder: ContextBuilder) {
        builder.addImplObject(PropertiesSource::class.java, YamlPropertiesSource(propertiesPath))
    }
}
